### PR TITLE
Add face recognition demo

### PR DESCRIPTION
**Pull request by elenash**

### CODE REVIEW

**Comment by elenash** [comment]
@snosov1 Please, review

File: demos/python_demos/face_recognition_demo/face_recognition_demo.py [hunk 1]
```diff
@@ -0,0 +1,1056 @@
...
+            label = osp.splitext(osp.basename(path))[0]
+            image = cv2.imread(path, flags=cv2.IMREAD_COLOR)
+
+            assert len(image.shape) == 3, \
+                "Expected an input image in (H, W, C) format"
+            assert image.shape[2] in [3, 4], \
+                "Expected BGR or BGRA input"
+
+            orig_image = image.copy()
+            if image.shape[0] == 4: # assume BGRA
```
Comment (zhiltsov-max) [inline demos/python_demos/face_recognition_demo/face_recognition_demo.py]: Wrong shape index here. We can either use 2 (as above) or, probably, just remove the line.

File: demos/python_demos/face_recognition_demo/face_recognition_demo.py [hunk 2]
```diff
@@ -0,0 +1,1056 @@
...
+        return model
+
+    def process(self, frame):
+        assert len(frame.shape) == 3, \
+            "Expected input frame in (H, W, C) format"
+        assert frame.shape[2] in [3, 4], \
+            "Expected BGR or BGRA input"
+
+        orig_image = frame.copy()
+        if frame.shape[0] == 4: # assume BGRA
```
Comment (zhiltsov-max) [inline demos/python_demos/face_recognition_demo/face_recognition_demo.py]: Wrong index.

**Comment by yssemaev** [comment]
Hi, per branching strategy, please merge to "develop" branch (not "master")

**Comment by snosov1** [comment]
@AlexanderDokuchaev please, review as well

**Comment by elenash** [comment]
> Hi, per branching strategy, please merge to "develop" branch (not "master")

This code is working with R5 and R1, and as far as I understand branching strategy, should be merged to master.

File: demos/python_demos/face_recognition_demo/face_recognition_demo.py [hunk 3]
```diff
@@ -0,0 +1,1056 @@
...
+        return [out[self.output_blob].flatten() for out in self.get_outputs()]
+
+    def _normalize(self, array, axis):
+        mean = array.mean(axis=axis)
+        array -= mean
+        std = array.std()
+        array /= std
+        return mean, std
+
+    def _get_transform(self, src, dst):
```
[the remark this answers is not in the document]
Comment (elenash) [reply demos/python_demos/face_recognition_demo/face_recognition_demo.py]: Not sure it's good idea, these funcs are related to landmarcs, not identifier results

**Comment by elenash** [comment]
@Wovchena @IRDonch Please, review and finally merge if no comments

**Comment by kchechil** [comment]
Jenkins please retry a build.

**Comment by kchechil** [comment]
My suggestion is to re-target to opencv:develop branch due to a lots of changes done.

**Comment by elenash** [comment]
> My suggestion is to re-target to opencv:develop branch due to a lots of changes done.

Done, can be reviewed

File: demos/python_demos/face_recognition_demo/face_detector.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        frame_width = frame.shape[-1]
+        frame_height = frame.shape[-2]
+
+        results = []
+        for output in outputs[0][0]:
+            result = FaceDetector.Result(output)
+            if result.confidence < self.confidence_threshold:
+                break # results are sorted by confidence decrease
+
+            result.clip(1, 1)
```
[the remark this answers is not in the document]
Comment (zhiltsov-max) [reply demos/python_demos/face_recognition_demo/face_detector.py]: These are the upper bounds for the vector range [(0, 0); (arg1, arg2)], a network theoretically can return something beyond this range.

[the remark this answers is not in the document]
Comment (Wovchena) [reply demos/python_demos/face_recognition_demo/face_detector.py]: There is `result.clip(frame_width, frame_height)` that does the same after rescale.

File: demos/python_demos/face_recognition_demo/face_recognition_demo.py [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+import numpy as np
+
+from openvino.inference_engine import IENetwork
+from ie_module import InferenceContext
+from landmarks_detector import LandmarksDetector
+from face_detector import FaceDetector
+from faces_database import FacesDatabase
+from face_identifier import FaceIdentifier
+
+DEVICE_KINDS = ['CPU', 'GPU', 'FPGA', 'MYRIAD', 'HETERO']
```
Comment (Wovchena) [inline demos/python_demos/face_recognition_demo/face_recognition_demo.py]: The list can be extended with HDDL because networks should be working there.

**Comment by elenash** [comment]
@ Wovchena Hopefully fixed your comments :)

File: demos/python_demos/face_recognition_demo/face_recognition_demo.py [hunk 6]
```diff
@@ -390,11 +392,12 @@ def run(self, args):
...
-            stream = 0
-        else:
+        stream = path
+        try:
+            stream = int(path)
+        except ValueError:
             assert osp.isfile(path), "Input file '%s' not found" % (path)
-            stream = path
+            pass
         return cv2.VideoCapture(stream)
```
Comment (Wovchena) [inline demos/python_demos/face_recognition_demo/face_recognition_demo.py]: There is no need in `pass`.

Or the check of file existence can be removed. `VideoCapture` can work not only with numbers and local files. For example, via `rtsp` protocol. I think it's better to check if `VideoCapture` `isOpened`.

**Comment by zmaslova** [comment]
Jenkins please retry a build

**Comment by zmaslova** [comment]
Jenkins please retry a build

**Comment by elenash** [comment]
@IRDonch Please, merge

File: demos/python_demos/face_recognition_demo/README.md [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+In particular, this demo uses 3 models to build a pipeline able to detect
+faces on videos, their keypoints (aka "landmarks"),
+and recognize persons using the provided faces database (the gallery).
+The following pretrained models can be used:
+
+* `face-detection-retail-0004` and `face-detection-adas-0001`, to detect faces and predict their bounding boxes;
+* `landmarks-regression-retail-0009`, to predict face keypoints;
+* `face-reidentification-retail-0095`, to recognize persons.
+
+For more information about the pre-trained models, refer to the [Open Model Zoo](https://github.com/opencv/open_model_zoo/tree/master/intel_models/index.md) repository on GitHub*.
```
Comment (IRDonch) [inline demos/python_demos/face_recognition_demo/README.md]: This should be changed to a relative link. Check the other demos for examples.

Comment (elenash) [reply demos/python_demos/face_recognition_demo/README.md]: done

**Pull request merged**